instant-distance/examples/colors.rs

```diff
@@ -17,6 +17,16 @@ fn main() {
 #[derive(Clone, Copy, Debug)]
 struct Point(isize, isize, isize);
 
+impl HasF32s for Point {
+    fn as_slice(&self) -> &[f32] {
+        unimplemented!("this will not work!")
```

Comment:
This is a good indication that this is the wrong level of abstraction. The `Hnsw` should not care what the type of `Point` is and, by extension, should not care what the inner type used for `Point` is. Instead, we could use an associated type in the `Point` trait (maybe `type Element` or `type Inner`) instead. (`STRIDE` should probably be an associated const for `Point`.)

Reply:
I did try a trait with an associated type `Element` (which had an associated const for STRIDE as well) for a while but that made resolving down to an f32 (or whatever type `Element` was) much harder to work with trait bounds for the SIMD implementation. I instead opted to, in the current iteration, reduce this down to just f32, since our internal types `Item, MapItem` were effectively limiting us to that anyway (due to our use of `OrderedFloat`). If you see a way where I could have a purely generic `Element` here, I'd prefer that as well.

Reply:
`Item` and `MapItem` apply `OrderedFloat` to the `distance` -- it's fine for `Point::distance()` to still yield `f32`, but I don't think that needs to constrain the data internal to the points?